lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp

```diff
@@ -1027,6 +1053,124 @@ SILCombiner::propagateConcreteTypeOfInitExistential(FullApplySite AI) {
   return propagateConcreteTypeOfInitExistential(AI, PD, PropagateIntoOperand);
 }
 
+/// A peephole optimizer that propagates concrete types of all arguments to
+/// Apply instruction. It handles all the arguments of an apply including self.
+SILInstruction *
+SILCombiner::propagateConcreteTypeOfInitExistentialsToAllApplyArgs(
+    FullApplySite AI) {
+  /// Check if legal to perform propagation.
+  if (!AI.hasSubstitutions())
+    return nullptr;
+  auto *Callee = AI.getReferencedFunction();
+  if (!Callee || !Callee->shouldOptimize() || Callee->empty())
```

Comment:
This restricts the optimization unnecessarily. It should work even if the callee of the apply is not a static `function_ref`. You should be looking at the operand of the apply's `SILFunctionType`, and walk the parameters of the function type. The arguments of the callee should not be needed.